### ISSUE TITLE
workflow for rotating generated intermediates without downtime?

### ISSUE THREAD

**Issue by slingamn** [hashicorp/vault#2581]
I'd like to use `pki/intermediate/generate` (with `type` set to `internal`) to generate intermediate private keys that are never extracted from Vault. The problem is that I can't see a way to rotate the private keys without downtime, since after each invocation of `pki/intermediate/generate`, the existing CA certificate is cleared and `pki/issue` is unavailable until the CSR has been signed and submitted via `pki/intermediate/set-signed`. Depending on how intermediates are being signed, this could be on the order of minutes or days. (In contrast, `pki/config/ca` can be used to atomically update the CA certificate and its private key, without downtime.)

A suggestion: we could add a type to the `generate` endpoint named `internal_pending`. This would generate the new private key, but store it in a distinct path in the storage backend, without overwriting the existing CA. Then `pki/intermediate/set-signed` would atomically install the pending key along with its signed CA certificate.

Thoughts? Thanks for your time.

**Comment by chrishoffman** [comment]
One way you could accomplish this is to run two separate intermediate CAs at different paths and rotate between them.  That way you can take as long as you need to sign your CSR and switch to the new one once it is ready.

**Comment by slingamn** [comment]
This seems like an awkward extra moving part (clients have to be told to retry failures from one backend against the other, or alternately some mechanism must be added for clients to dynamically discover the currently working endpoint).

**Comment by jefferai** [comment]
Whether it's awkward depends entirely on your setup. If you use Consul, just pop the current valid path in a K/V entry and have clients use that value.

If you're using Vault for your root, the downtime should be on the order of seconds; if you're not, you can always generate an intermediate, get it signed, then use `pki/config/ca` to set both at once.

**Comment by jefferai** [comment]
BTW, the pending idea is fine, but another way to go would be to reuse the key and be able to pull out a new CSR to be signed, say if the `type` is `csronly` or so. That's probably more of the "correct" way to do it.

**Comment by slingamn** [comment]
Yeah, installation via `pki/config/ca` is what I'm doing now.

The main goal is to rotate the private key of the intermediate, so I'd like to generate a new private key, rather than generating a new CSR for the old one. I like `csronly` as a name, though.

**Comment by jstachowiak** [comment]
It would be nice to have an official guidance on how to renew an issuing CA certificate with *an existing key pair* and *a new key pair* when using PKI with Vault.

To renew an issuing CA certificate with an existing key pair, you need to store CSR somewhere as currently Vault's HTTP API only provides the CSR when you generate an intermediate (assuming you don't want to export the private key). This scenario is helpful when you renew an issuing CA certificate at its half-life to facilitate issuing certificates for their full life. In other words, issuing CA certificate must be valid at least twice the lifetime of the issued certificates.

If you want to renew with a new key pair, then in fact you need to mount PKI secrets engine at a different path because clients using issued certificates may rely on paths for things like:
- Authority Information Access extension which contains a URL  to download the issuing CA certificate
- CRL Distribution Points extension which provides the location of signed CRLs
- more?

We store which path is active at any given time in Consul. I guess both workflows could be simplified if supported through Vault's HTTP API.

**Comment by cipherboy** [comment]
As I alluded to in #11024 and #10815, the new multi-issuer feature in https://github.com/hashicorp/vault/pull/15277 and doc'd in https://github.com/hashicorp/vault/pull/15238 should address this. This feature was designed with rotation in mind. :-) Feel free to test it out.
